### ISSUE TITLE
La configuration des variables est écrasée en cas d'erreur mysql

### ISSUE THREAD

**Issue by denis-dordoigne** [AlternC/AlternC#143]
Les paramètres d'entrée de la fonction variable_get sont les suivants :

```
 * Return a persistent variable.
 *
 * @param $name
 *   The name of the variable to return.
 * @param $default
 *   The default value to use if this variable has never been set.
 * @param $createit_comment 
 *   If variable doesn't exist, create it with the default value
 *   and createit_comment value as comment
```

Avoir une fonction _get_ qui peut faite un _set_ est déjà contre-intuitif, mais en plus le comportement n'est pas exactement celui documenté, puisque la fonction :
* tente de récupérer la valeur de la variable $name
* si elle n'arrive pas à la récupérer, quelle que soit la raison, et que $createit_comment est défini, la fonction appelle variable_set qui elle :
  * crée la variable en base si elle n'existe pas
  * modifie la valeur de la variable si elle existe

Résultat : si le SELECT pour récupérer la variable échoue pour une raison technique (perte de connexion, timeout, etc.), la variable se retrouvera écrasée (si le problème technique a disparu entretemps) !

Pistes pour résoudre le bug :
* différencier dans les appels en base les erreurs mysql (qui générerait une exception) de l'absence de résultat
* ajouter un argument "update_if_exist" à variable_set, qui pourrait être vraie par défaut, mais qui serait positionnée à faux en cas d'appel par variable_get

**Comment by lelutin** [comment]
nous avons vu exactement ce problème chez Koumbit quand le serveur de base de données étaient trop engorgé. à cause des variables qui disparaissaient, plusieurs fonctionnalités comme piwik cessaient de fonctionner, et il fallait alors restaurer la table des variables d'un backup.

notre solution pour stopper le problème a été de mettre en commentaire l'appel à variable_set dans variable_get, mais ça brise alternc.install.

avoir une fonction _get qui fait un _set en cas d'erreur est un cas parfait d'exemple d'anti-pattern en développement logiciel.

si le seul cas qui devrait être réglé par cet appel à variable_set c'est pour que alternc.install fonctionne bien, c'est alors dans ce script qu'il faudrait vérifier le code d'erreur et appeler variable_set si la variable n'existe pas. et non *pas* pendant tous les appels normaux.

**Comment by denis-dordoigne** [comment]
En attendant que ce soit corrigé, j'ai bidouillé la base de données pour éviter de perdre des données. Lorsque le variable_get fait un variable_set, il tente de faire un update du commentaire associé à la variable, j'ai profité de ce comportement pour faire un trigger protecteur.

Étape 1 : modifier le commentaire de toutes les variables qu'on ne veut pas voir écrasées. Il faut modifier le champ _comment_ de la table _variable_ ; pour_ ma part je ne me suis pas fatigué, j'ai juste fait : 
`update variable set comment=concat(comment, ' (protected)');`

Étape 2 : créer un trigger qui écrase les updates lorsque le commentaire est modifié en même temps que la valeur : 
```
delimiter //
CREATE TRIGGER alternc_issue143 BEFORE UPDATE ON variable FOR EACH ROW
BEGIN
  IF (NEW.`comment` <> OLD.`comment`) THEN
    SET NEW.value = OLD.value;
    SET NEW.`comment` = OLD.`comment`;
  END IF;
END;//
delimiter ;
```

Résultat : 
- variable_get ne peut plus mettre à jour une variable dans son usage actuel
- on peut toujours modifier une variable depuis le panneau administrateur

**Comment by vincib** [comment]
Suggestion : que la page du panel *s'arrête* net et sorte une erreur en cas d'erreur MySQL (via notre classe utilisant PDO, c'est simple ;) )

**Comment by vincib** [comment]
dans le __construct de pdo je propose que l'on se connecte ainsi : 
          $this->pdo_instance = new PDO($dsn, $user, $passwd, 
array(PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION ));

l'ajout du array() d'options fait qu'une page ou un script PHP sera arrêté NET en cas d'erreur MySQL.
(on a ça sur la prod mutu Octopuce depuis un mois, pour l'instant sans dommage ;) ) 
à tester lourdement par contre ...

**Comment by lelutin** [comment]
@vincib si je comprends bien, l'appel à variable_get() est là pour créer les valeurs par défaut pendant l'installation (wtf! c'est l'installeur qui devrait faire ça)

donc c'est un cas d'usage qu'il faut vérifier. je crois que c'est le seul cas "valide" en place

**Comment by lelutin** [comment]
je suis d'accord avec le stop NET par contre. ça devrait être le comportement souhaité s'il y a un trouble d'accès à la base de données.

donc la solution que je propose à mon commentaire précédent c'est de déplacer la création initiale des variables dans alternc.install et dans la fonction variable_get() remplacer l'appel à variable_set() par un lancement d'erreur.

**Comment by camlafit** [comment]
Hello

En complément, je pense qu'il faut faire évoluer le code avec une variable_update() en plus

* variable_set() définit une variable inexistante , et return si cela a été inséré
* variable_get() retourne la valeur ou la valeur défaut demandé ou false si rien n'est trouvé
* variable_update() met à jour une variable ou la créé si absente

Cela ne devrait pas changer les prototype des fonctions et on évite le mélange des genres avec un get qui fait du set
